### ISSUE TITLE
Patching regression in 0.12.6

### ISSUE THREAD

**Issue by sashabaranov** [bokeh/bokeh#6536]
Here is the screenshot of 1D histogram, drawn with `multi_line` and `y_axis_type="log"` in bokeh==0.12.4. Background was done using `plot.patch`

![screen shot 2017-06-26 at 12 20 45](https://user-images.githubusercontent.com/677093/27533709-3e20e5b0-5a6d-11e7-8a73-56e9f242d51d.png)

Same plot using 0.12.6:

![screen shot 2017-06-26 at 12 23 16](https://user-images.githubusercontent.com/677093/27533795-8ba9d5a8-5a6d-11e7-8ff0-a51d62c430cc.png)

Also in 0.12.6, `.patch` is used for plotting blue background

![screen shot 2017-06-26 at 12 27 27](https://user-images.githubusercontent.com/677093/27533824-a3c4524e-5a6d-11e7-9340-5bc0a195af74.png)

**Comment by sashabaranov** [comment]
I guess it is somehow connected to #6285

**Comment by bryevdv** [comment]
A minimal complete example is needed to investigate

**Comment by sashabaranov** [comment]
@bryevdv here you go

```python
from bokeh.io import show
from bokeh.plotting import figure
from bokeh.models import Range1d

data = {
    'binning': [(0, 1), (1,2), (3,4), (5,6)],
    'values': [100, 80, 60, 20]
}


plot = figure(
    x_range=(data['binning'][0][0], data['binning'][-1][-1]),
    tools="pan,wheel_zoom,box_zoom,reset",
    active_drag="box_zoom",
    toolbar_location="right",
    x_axis_type="auto",
    y_axis_type="log"
)


xs = []
ys = []


xs.append((data['binning'][0][0], data['binning'][0][0]))
ys.append((0, data['values'][0]))

for i in xrange(len(data['values'])):
    x = data['binning'][i]
    y = data['values'][i]

    if i == len(data['values']) - 1:
        next_y = 0
    else:
        next_y = data['values'][i+1]

    # horizontal bar
    xs.append((x[0], x[1]))
    ys.append((y, y))

    # vertical bar
    xs.append((x[1], x[1]))
    ys.append((y, next_y))

plot.multi_line(xs, ys, color="blue", line_width=2)


# Background
p_xs = [item for sublist in xs for item in sublist]
p_ys = [item for sublist in ys for item in sublist]

plot.patch(p_xs, p_ys, alpha=0.6, line_width=0, fill_color="red")

show(plot)
```

**Comment by bryevdv** [comment]
@sashabaranov perfect thanks, FYI that other PR regards the `.patch` protocol for data sources, not the `.patch` glyph, I doubt that is related to this issue.

**Comment by bryevdv** [comment]
This is almost certainly related to handling of log axes. As you may know, zero values are problematic on log scales. If I replace `next_y = 0` with `next_y = 10e-10`, then the output is substantially improved. It's not quite perfect but that's only due to a missing y value for `x=0`. 

Given this workaround, I am pushing this back to `0.12.8` since the current milestone is already overly optimistic in terms of what can be accomplished.  

ping @clairetang6 for any quick thoughts.

**Comment by sashabaranov** [comment]
Here is a data sample where this workaround does not really help:

```
data = {
    'binning': [(0, 1), (1,2), (3,4), (5,6), (6,7), (7,8), (8,9), (9, 10)],
    'values': [0, 4, 20, 10, 0, 5, 0]
}
```

**Comment by sashabaranov** [comment]
Okay, found good workaround:
1. Set `plot.y_range.start` to min nonzero value / 2
2. Check whether `y` or `next_y` is close to 0
3. If yes, set it to `plot.y_range.start`

**Comment by clairetang6** [comment]
I would need to think a little more about whether the special case for 0 would be worth it. It definitely is not the technically right thing to do. It seems to me like the previous behavior was caused by a bug that was mapping 0's to the x-axis.. 

The workaround posted by @sashabaranov looks to solve the problem. Making this plot just requires you to be explicit about where the lowest values should be drawn (`plot.y_range.start`).

**Comment by bryevdv** [comment]
>  It definitely is not the technically right thing to do. 

Ehhhhhh sort of :) It strikes me that e.g bar charts on a log scale should be reasonably constructible, in principle.

**Comment by clairetang6** [comment]
Right, for bar charts and this histogram, it makes a little bit of sense (though there is a visualization error (error may be too harsh a word) in the same way for bar charts that don't start at 0, where the lengths of the bars aren't proportional to the values). 

But in default mapping 0 to x-axis, you have the problem that the x-axis does occur at some y value. So it would be unclear whether something rendered at the x-axis was meant to be a 0 or that y value (probably not a huge issue).

The main thing to consider would be if there is another possible intention for 0 in the log setting, and I can't think of one besides using 0 instead of NaN.

**Comment by erickee** [comment]
Numpy handles 0 values in log axes properly. They map to -inf, which has the following implications
1. Line plot example:  
    x = [1, 2, 3], y = [1, 0, 2] 
would become
    x = [1, 1, 3, 3], y = [1, -ymin, -ymin, 2] 
where ymin is a value that must always be below the bottom of the y axis limits, ideally even when zooming out. 

This is really valuable for plotting probability distributions and histograms that have zero bins, and I hope Bokeh will match something like the behavior of numpy/pyplot.

**Comment by bryevdv** [comment]
Copying from comment on closed #8343 

> I'd propose a solution that allows scales to be configured with a policy that describes what to do with special values like inf and -inf:
> ```
> p.x_scale.positive_inf = "zero"       # or "screen_max" or None (default)
> p.y_scale.negative_inf = "screen_min" # or "zero" or None (default)
> ```
> i.e. "zero" means map the +/-inf to 0 and "screen_max" means map +inf to whatever data value corresponds to the current range max. If a policy is None that corresponds to the current behaviour.

This is basically what @erickee describes above. This should be accessible, with help and guidance to anyone with even a modicum of JS experience. I would be more than happy to help advise anyone interested in working on this sooner than the core devs are able to get to it.

**Comment by bryevdv** [comment]
To elaborate on the above idea, the gist is to allow users to specify what log scales should do when inf values are computed. So for `y_scale.negative_inf `, which would control what y_scale should do for negative inf values, the options are:

* `None`  - do nothing / current behavior
* `"zero"` - always return 0 unconditionally
* "screen_min" - return the current value of `y_range.min`

So `Scale.compute` and `Scale.v_compute` would look at these policies before returning their values, and adjust them accordingly if the value would be inf.  

The "zero" policy is most useful for making bars or filled areas work. 

The "screen_min" policy simulates a "real -inf", e.g. when the plot is panned. This would be useful for line plots, if you pan down, the line always maps to the lowest visual point on the plot no matter how far you pan. 

cc @birdsarah

**Comment by bryevdv** [comment]
I realized that my and @erickee's  are somewhat different. I missed that the values to plot in his case have extra values inserted to make a line appear to go "straight down" on both sides of a -inf value. That could only happen in the renderer, and every glyph would have to be customized how to handle the situation. I was trying to imagine a way to handle the issue in one place, by manipulating the scale's behaviour. Now I'm not sure, maybe this requires more thought.

**Comment by majmin7** [comment]
Bryan, the behavior in which additional values are inserted is essential to handle -inf correctly. (Matlab behaves that way if I recall correctly.) A hacky solution would map -inf to some very large number relative to the next smallest value, and then set the axis range to the next smallest value. That would lead to numerical issues in some cases though, so a general solution is almost certainly required. This issue is fundamental to the concept of +/- inf for the same reason that the moon appears to be in the same direction regardless of how fast you drive down the highway.

**Comment by erickee** [comment]
Bryan, the behavior in which additional values are inserted is essential to handle -inf correctly. (Matlab behaves that way if I recall correctly.) A hacky solution would map -inf to some very large number relative to the next smallest value, and then set the axis range to the next smallest value. That would lead to numerical issues in some cases though, so a general solution is almost certainly required. This issue is fundamental to the concept of +/- inf for the same reason that the moon appears to be in the same direction regardless of how fast you drive down the highway.

**Comment by birdsarah** [comment]
Notes / questions for myself:
  * How do I simulate an example that uses `models/scales/log_scale.ts::LogScale.compute()` rather than  `models/scales/log_scale.ts::LogScale.v_compute()`?

**Comment by bryevdv** [comment]
* I am not sure anything uses (single value) compute, it's there for completeness but I can't say if anything uses it for sure. 

* IIRC in our usage source is data space and target is screen

**Comment by michaelaye** [comment]
Didn't the discussion here go far too far? Why did nobody mention whatever MPL does in this case? I can think of a million use cases where a bunch of numbers don't cover all the x-axis but have so vastly differing frequency that a log is needed to see the low numbers at all, e.g. here for all exposure times of the MRO HiRISE camera: 
![image](https://github.com/bokeh/bokeh/assets/69774/9a666992-6653-4636-89a9-c2dc4fdb73a5)
As one can see, no issue for MPL.
IMHO (maybe I'm missing something), this is something so simple, bokeh should be able to decide what to do here, for the sake of the end users.

**Comment by birdsarah** [comment]
Can I be unassigned from this issue?
